### ISSUE TITLE
Support speculative decoding in `server` example

### ISSUE THREAD

**Issue by mscheong01** [ggerganov/llama.cpp#5877]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [x] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [x] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

provide speculative decoding through `server` example.

# Motivation

Noticed this topic popped up in several comments ([1](https://github.com/ggerganov/llama.cpp/issues/4286#issuecomment-1839494720), [2](https://github.com/ggerganov/llama.cpp/issues/4216#issuecomment-1826380687), [3](https://github.com/ggerganov/llama.cpp/issues/4216#issuecomment-1970810228)) but it seems we haven't officially opened an issue for it. I'm creating this to provide a space for focused discussion on how we can implement this feature and actually get this started.

# Possible Implementation

perhaps move speculative sampling implementation to `common` or `sampling`?

**Comment by vietanh125** [comment]
Any updates for this?

**Comment by mscheong01** [comment]
@vietanh125 Not yet, but contributions are welcome 😃

**Comment by ggerganov** [comment]
There is ongoing related work in https://github.com/ggerganov/llama.cpp/pull/6828. Though I haven't had time to look in details yet

**Comment by kerthcet** [comment]
Sorry, does that means the server doesn't support speculative decoding? However, I can run it with commands like below in Kubernetes

Just a sample:
```
spec:
  containers:
  - args:
    - -m
    - /workspace/models/llama-2-7b.Q8_0.gguf
    - -md
    - /workspace/models/llama-2-7b.Q2_K.gguf
    - --port
    - "8080"
    - --host
    - 0.0.0.0
    - -fa
    command:
    - ./llama-server
```

**Comment by ggerganov** [comment]
Not yet supported

**Comment by kerthcet** [comment]
Ok so the `-md` doesn't work here 😀

**Comment by theo77186** [comment]
The #6828 PR is a distinct technique that uses a lookup file to speculate tokens instead of using a draft model, there seems to have less speedup than draft-based speculative decoding.

**Comment by Hoernchen** [comment]
Support would be really nice to have because now there is the offical llama 3.2 in 1b and 3b which should be suitable for 8/70b 3.1, at least according to the offical HF notebook: https://github.com/huggingface/huggingface-llama-recipes/blob/main/assisted_decoding_8B_1B.ipynb

**Comment by enn-nafnlaus** [comment]
Wait, what happened?  I used to run llama-server with speculative decoding with -md.  I just "upgraded" and -md went away.  now there's a separate program called llama-speculative, but doesn't appear to be a server.  Sigh  :(  Guess I have to downgrade and find the version where it went away....

**Comment by theo77186** [comment]
The last commit with `-md` in llama-server was https://github.com/ggerganov/llama.cpp/commit/554c247caffed64465f372661f2826640cb10430 but it never worked anyway. The speculative decoding flags were silently discarded and no speculator model was loaded.

**Comment by sammcj** [comment]
Came to ask the same as other folks have stated here - looks like `-md` is no longer an option for the server. @ggerganov do you have any plans to implement speculative decoding for the server component?

**Comment by oxfighterjet** [comment]
Is anyone working on this issue? Or is this possibly blocked by something?

I am already preparing for this feature to be implemented in Ollama, but depend on this feature being implemented in llama-server here.

I don't mind giving this issue here a shot, it is labeled as good first issue and if that's true would make it suitable for my first commit.

I had a quick look and from what I see there is already an example of implementation in speculative. I assume I can use that as a hint for implementing it at the server level. 

Are there any additional pointers or specific considerations for the implementation I should be aware of?

**Comment by ggerganov** [comment]
At the very least the `llama-speculative` example has to be fixed first (https://github.com/ggerganov/llama.cpp/issues/10176#issuecomment-2459450448) and then demonstrate some meaningful gains from having this feature implemented in the server.

**Comment by sammcj** [comment]
> ~53% performance increase when using the draft model

Glad to hear this, this is pretty similar to ExllamaV2.

The Qwen 2.5 model family is a good example for this as well, you can basically use the small 1.5b or even 0.5b model for the draft with the big 72b model and get an excellent boost.

**Comment by oxfighterjet** [comment]
I also ran some smaller scale tests, which I wanted to share to bring some additional perspective (this is on an RTX 3060 with 12GB of VRAM, so can't find as large models):

`./llama-cli -m llama3.1\:8b-instruct-q8_0 -p "I believe, in one sentence, that the meaning of life is" -ngl 33`

```
llama_perf_sampler_print:    sampling time =     601.84 ms /   913 runs   (    0.66 ms per token,  1517.01 tokens per second)
llama_perf_context_print:        load time =    3575.03 ms
llama_perf_context_print: prompt eval time =      56.80 ms /    14 tokens (    4.06 ms per token,   246.46 tokens per second)
llama_perf_context_print:        eval time =   27636.04 ms /   898 runs   (   30.78 ms per token,    32.49 tokens per second)
llama_perf_context_print:       total time =   29298.72 ms /   912 tokens
```

`./llama-speculative -m llama3.1\:8b-instruct-q8_0 -ngl 33 -md llama3.2\:1b-instruct-q8_0 -ngld 17 -p "I believe, in one sentence, that the meaning of life is"`

```
encoded   14 tokens in    0.054 seconds, speed:  258.250 t/s
decoded  673 tokens in   12.132 seconds, speed:   55.473 t/s

n_draft   = 5
n_predict = 673
n_drafted = 730
n_accept  = 526
accept    = 72.055%

draft:

llama_perf_context_print:        load time =    2130.39 ms
llama_perf_context_print: prompt eval time =    7303.87 ms /   305 tokens (   23.95 ms per token,    41.76 tokens per second)
llama_perf_context_print:        eval time =    3685.16 ms /   584 runs   (    6.31 ms per token,   158.47 tokens per second)
llama_perf_context_print:       total time =   12191.61 ms /   889 tokens

target:

llama_perf_sampler_print:    sampling time =     466.15 ms /   673 runs   (    0.69 ms per token,  1443.75 tokens per second)
llama_perf_context_print:        load time =    3573.26 ms
llama_perf_context_print: prompt eval time =    4898.77 ms /   890 tokens (    5.50 ms per token,   181.68 tokens per second)
llama_perf_context_print:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:       total time =   14322.06 ms /   891 tokens
```

Which, if I read that correctly, bumps the speed from `32.49` t/s to `55.473` t/s, hence a speedup of 70% (excluding model loading times).

**Comment by nullnuller** [comment]
Any progress on allowing speculative decoding in the server?

**Comment by ggerganov** [comment]
It's already supported - this issue hasn't been closed.

https://github.com/ggerganov/llama.cpp/pull/10455
